### DISCUSSION TITLE
Feature Request: Handle parser failure per query statement instead of per query string

### DISCUSSION THREAD

**Discussion by stephaniewang526** [duckdb/duckdb#7854] [Ideas]
Hi, 
The current code will thrown an exception if an entire sql string including strings/query containing multiple statements fail to parse and DuckDB passes on to other parser extensions to handle what happens next. This creates complexity when a valid DuckDB statement and a custom statement appear in the same multi-statment query. For example:
```
con.Query("CREATE DATABASE main_db; USE main_db")
```
(where `CREATE DATABASE` is not a supported in DDB but USE is supported). This means that the our parser would also need to parse DDB-parseable statements which is not ideal. 

Ideally, our parser extension parses what DDB does not parse and we rely on DDB to parse statements that it parses.

Wonder if it's possible to change the exception throwing logic [here](https://github.com/duckdb/duckdb/blob/master/src/parser/parser.cpp#L156-L172) so that it will check by statement?

**Comment by Tishj** [comment]
Looking at the code, we rely on the parser to let us know the boundaries between statements.
Only here: `transformer.TransformParseTree(parser.parse_tree, statements);` is the parse tree transformed to statements, which only happens if the parser was successful.

I think the parser error contains the index at which the error occurred, if the parse tree is valid up to that point we might be able to extract the statements leading up to it, then from that point on look for the next statement boundary, and only feed the extension the statement that we failed to parse.

**Reply by Tishj** [reply]
I'm gonna try a pass at fixing this

EDIT:
On further inspection, we likely need to take a closer look at the problem

**Reply by stephaniewang526** [reply]
Thanks. Please keep me posted.

**Reply by stephaniewang526** [reply]
Fixed in https://github.com/duckdb/duckdb/pull/7868
